chef/cookbooks/crowbar/recipes/prepare-os-upgrade.rb

```diff
@@ -0,0 +1,65 @@
+# Cookbook Name:: crowbar
+# Recipe:: prepare-os-upgrade
+#
+# Copyright 2013-2016, SUSE LINUX Products GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This recipe prepares a script which when executed (at selected time)
+# 1. removes old repositories
+# 2. adds correct new ones
+# 3. runs zypper dup to upgrade the node
+
+arch = node[:kernel][:machine]
+
+old_repos = Provisioner::Repositories.get_repos(
+  node[:platform], node[:platform_version], arch
+)
+
+target_platform, target_platform_version = node[:target_platform].split("-")
+new_repos = Provisioner::Repositories.get_repos(
+  target_platform, target_platform_version, arch
+)
+
+# Find out the location of the base system repository
+provisioner = search(:node, "roles:provisioner-server")[0]
+admin_ip = Chef::Recipe::Barclamp::Inventory.get_network_by_type(provisioner, "admin").address
+web_port = provisioner[:provisioner][:web_port]
+provisioner_web = "http://#{admin_ip}:#{web_port}"
+
+web_path = "#{provisioner_web}/#{node[:platform]}-#{node[:platform_version]}/#{arch}"
+old_install_url = "#{web_path}/install"
+
+web_path = "#{provisioner_web}/#{node[:target_platform]}/#{arch}"
+new_install_url = "#{web_path}/install"
+
+# try to create an alias for new base repo
+zypper_out = `zypper lr --details | grep #{old_install_url}`
+repo_alias = zypper_out.split("| ")[1] || "SLES12-SP1-12.1-0"
+new_alias = repo_alias.gsub("SP1", "SP2").gsub(node[:platform_version], target_platform_version)
+
+template "/usr/local/bin/upgrade_cloud_node.sh" do
```

Comment:
I'd put in `/usr/sbin` (not `/usr/local` because it's not created by the user, and not `bin` because it's really very special). Also, can we do s/_/-/ in the filename.

Actually, I'd rename it to /usr/sbin/crowbar-upgrade-os. (No mention of cloud, as it's not cloud specific)